database/storage/bbolt/bbolt.go

```diff
@@ -264,33 +271,33 @@ func (b *BBolt) MaintainRecordStates(ctx context.Context, purgeDeletedBefore tim
 			// check if we need to do maintenance
 			meta := wrapper.Meta()
 			switch {
-			case meta.Deleted > 0 && meta.Deleted < purgeThreshold:
+			case meta.Deleted == 0 && meta.Expires > 0 && meta.Expires < now:
+				if shadowDelete {
+					// mark as deleted
+					meta.Deleted = meta.Expires
+					deleted, err := wrapper.MarshalRecord(wrapper)
+					if err != nil {
+						return err
+					}
+					err = bucket.Put(key, deleted)
+					if err != nil {
+						return err
+					}
+
+					// reposition cursor
+					c.Seek(key)
```

Comment:
There's no need to reposition the cursor if we just mark the record as deleted (i.e. shadowDelete).